server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -125,30 +134,71 @@ static void executeInternal(SearchContext searchContext) throws QueryPhaseExecut
                 }
             }
 
-            final LinkedList<QueryCollectorContext> collectors = new LinkedList<>();
-            // whether the chain contains a collector that filters documents
-            boolean hasFilterCollector = false;
+            // create the top docs collector last when the other collectors are known
+            final TopDocsCollectorContext topDocsFactory = createTopDocsCollectorContext(
```

Comment:
Should we rename `TopDocsCollectorContext` to `TopDocsCollectorFactory`?

Reply:
You actually called the variable `topDocsFactory` :)

Reply:
I will keep this for a small follow-up if you don't mind.

Reply:
sure